### ISSUE TITLE
spaced-comment can't be used only for single line comments

### ISSUE THREAD

**Issue by noppa** [eslint/eslint#9898]
**What rule do you want to change?**
[spaced-comment](https://eslint.org/docs/rules/spaced-comment)

**Does this change cause the rule to produce more or fewer warnings?**
Fewer.

**How will the change be implemented? (New option, new default behavior, etc.)?**
The "block"/"line" configuration options that are currently supported would be augmented to allow new string values `"always"`, `"never"` and `"ignore"`. `"always"` and `"never"` would correspond to the second argument of the `spaced-comment` rule. `"ignore"` would disable the rule completely for block/line comments.

Hence,
```json
"spaced-comment": ["error", "always", {
  "line": "never",
  "block": "never"
}]
```
would be equivalent to
```json
"spaced-comment": ["error", "never"]
```
and
```json
"spaced-comment": ["error", "always", {
  "line": "ignore",
  "block": "ignore"
}]
```
would be equivalent to
```json
"spaced-comment": 0
```
In practice, these new values would be used only in "line" **or** "block" configuration to tell the linter that these two types of comments are to be handled differently from each other.

Alternative or complementary way to describe this would be to add a new option for the "line"/"block" configuration objects. For example
```json

"spaced-comment": ["error", "always", {
    "line": {
        "markers": ["/"],
        "exceptions": ["-", "+"]
    },
    "block": {
        "spaces": "never",
        "markers": ["!"],
        "exceptions": ["*"],
        "balanced": true
    }
}]
```
Here, `block.spaces` is a new configuration option. With the above configuration, leading spaces would be required for line comments, but disallowed for block comments.

**Please provide some example code that this change will affect:**

```js
/* eslint spaced-comment: [2, "always", { "block": "ignore" }] */
// foo
function foo(a/*:string*/) {}
```

**What does the rule currently do for this code?**
Currently, `{ "block": "ignore" }` configuration is invalid so the linting fails. With a valid configuration  
`/* eslint spaced-comment: [2, "always"] */` The `/*:string*/` block comment in line 3 would give an error
> Expected space or tab after '/*' in comment.

**What will the rule do after it's changed?**
The configuration that is described above would be valid and no errors would be given. `/*:string*/` is allowed because the rule is ignored for block comments. `/* :string*/` wouldn't give an error either.

**Previous discussion:**
#2897

**Comment by platinumazure** [comment]
Hi @noppa, thanks for the issue.

It looks like the rule currently does not support enabling the rule for line comments and disabling (or using the opposite option) for blocks.

I think we should augment the rule schema to support different options for line/block. I think we could consider supporting both string values for block/line (e.g., `"block": "never"` or always or ignore), and we could also consider augmenting the object option to change the primary enforcement (e.g., `"block": { "spaces": "never", "exceptions": [], "markers": [] }`).

If you think that sounds like a reasonable approach, would you mind editing your post into a [Rule Change Proposal](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md) template with a few more examples? Thanks!

**Comment by noppa** [comment]
@platinumazure, I edited the post based on your suggestions. A few thoughts:
* Is it worth it to support the `"block": "never"` shorthand? If it's possible to flip the way comments are handled like this, then the more flexible syntax `"block": { "spaces": "never" }` is also going to be needed because one might want to add markers and exceptions. I think only supporting the more verbose configuration object would be ok and could save some work (maybe?).
* Wouldn't `"off"` instead of `"ignore"` be more aligned with the way rules are turned off with `0` or `"off"`?

**Comment by not-an-aardvark** [comment]
Thanks for your interest in improving ESLint. Unfortunately, it looks like this issue didn't get consensus from the team, so I'm closing it. We define [consensus](http://eslint.org/docs/maintainer-guide/issues.html#consensus) as having three :+1:s from team members, as well as a team member willing to champion the proposal. This is a high bar by design -- we can't realistically accept and maintain every feature request in the long term, so we only accept feature requests which are useful enough that there is consensus among the team that they're worth adding.

Since ESLint is pluggable and can load custom rules at runtime, the lack of consensus among the ESLint team doesn't need to be a blocker for you using this in your project, if you'd find it useful. It just means that you would need to [implement the rule](http://eslint.org/docs/developer-guide/working-with-rules) yourself, rather than using a bundled rule that is packaged with ESLint.
